### PR TITLE
Monorepo

### PR DESCRIPTION
**Pull request by bdehamer**
This is a pretty major refactoring of the `tuf-js` code:

* Re-organizes the repository into a monorepo hosting multiple npm packages.
* Splits the existing `tuf-js` package into two separate packages:
  * `@tufjs/models` containing all of the model objects representing the various TUF metadata types (everything which was previously found in the "/models" source directory.
  * `tuf-js` containing just the classes which implement the TUF client workflow (`Updater`, `Fetcher`, and `TrustedMetadataStore`). The models are imported from the new `@tufjs/models` package.
* Introduces a new `mock-repo` package which can be used to easily mock all of the TUF repository HTTP requests during testing.
* Re-organizes the code in the "examples" directory to work better with the repository layout.

This PR does NOT address the publishing workflow for these new packages. Will handle that in a separate change.

### CODE REVIEW

File: packages/models/src/base.ts [hunk 1]
```diff
@@ -18,6 +17,20 @@ export interface SignedOptions {
...
+  Snapshot = 'snapshot',
+  Targets = 'targets',
+}
+
+export function isMetadataKind(value: unknown): value is MetadataKind {
+  return (
+    typeof value === 'string' &&
+    Object.values(MetadataKind).includes(value as MetadataKind)
+  );
+}
```
Comment (bdehamer) [inline packages/models/src/base.ts:32]: Relocated this code to work around some circular import issues.

File: packages/models/src/metadata.ts [hunk 2]
```diff
@@ -52,6 +52,17 @@ export class Metadata<T extends MetadataType> implements Signable {
...
+  public sign(signer: (data: Buffer) => Signature, append = true): void {
+    const bytes = canonicalize(this.signed.toJSON());
+    const signature = signer(bytes);
+
+    if (!append) {
+      this.signatures = {};
+    }
+
+    this.signatures[signature.keyID] = signature;
+  }
```
Comment (bdehamer) [inline packages/models/src/metadata.ts:64]: New function to support the generation of signatures for metadata files. Needed by the new `repo-mock` package.

File: packages/models/src/root.ts [hunk 3]
```diff
@@ -51,7 +50,19 @@ export class Root extends Signed {
...
+    if (!this.roles[role]) {
+      throw new ValueError(`role ${role} does not exist`);
+    }
+
+    if (!this.roles[role].keyIDs.includes(key.keyID)) {
+      this.roles[role].keyIDs.push(key.keyID);
+    }
+
+    this.keys[key.keyID] = key;
+  }
```
Comment (bdehamer) [inline packages/models/src/root.ts:63]: New function to support the addition of new keys to the root. Needed by the new `repo-mock` package.

File: packages/models/src/targets.ts [hunk 4]
```diff
@@ -28,7 +27,11 @@ export class Targets extends Signed {
     this.delegations = options.delegations;
   }
 
-  public equals(other: Targets): boolean {
+  public addTarget(target: TargetFile): void {
+    this.targets[target.path] = target;
+  }
+
```
Comment (bdehamer) [inline packages/models/src/targets.ts:33]: New function to support the addition of targets to the targets-metadata. Needed by the new `repo-mock` package.

File: packages/models/src/utils/types.ts [hunk 5]
```diff
@@ -1,10 +1,3 @@
-export enum MetadataKind {
-  Root = 'root',
-  Timestamp = 'timestamp',
-  Snapshot = 'snapshot',
-  Targets = 'targets',
-}
```
Comment (bdehamer) [inline packages/models/src/utils/types.ts:6]: Relocated to "src/base.ts"

File: packages/repo-mock/src/__tests__/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    mockRepo(baseURL, [targetFile]);
+    const target = await fetch(`${baseURL}/targets/foo.txt`);
+    expect(target).toEqual(targetFile.content);
+
+    // Non-existent target
+    await expect(fetch(`${baseURL}/targets/bar.txt`)).rejects.toThrow(/404/);
+  });
+});
+
+async function fetch(url: string): Promise<string> {
```
Comment (ejahnGithub) [inline packages/repo-mock/src/__tests__/index.test.ts:76]: why do we need a custom fetch function?

File: packages/repo-mock/src/metadata.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    version: 1,
+    length: meta.length,
+    hashes: { sha256: digestSHA256(meta) },
+  });
+}
+
+function getExpires(): string {
+  const expires = new Date();
+  expires.setFullYear(expires.getFullYear() + 1);
+  return expires.toISOString();
```
Comment (ejahnGithub) [inline packages/repo-mock/src/metadata.ts:93]: In tuf spec, all metadata's EXPIRES are the same as is described for the root.json file. Just wonder by calling the function, would that have problem that there is a little offset between each EXPIRES data in metadata and root

**Review by ejahnGithub — APPROVED** [review]
Left some comments. Overall look good to me!

File: packages/repo-mock/src/metadata.ts (hunk 7 above)
Comment (bdehamer) [reply packages/repo-mock/src/metadata.ts:93]: I was just looking at the metadata files in the Sigstore TUF repo and they all have different expiration values . . . which sorta makes sense cause the root may be long-lived (expires 1 year from now) whereas the timestamp metadata should expire pretty quickly.

File: packages/repo-mock/src/__tests__/index.test.ts (hunk 6 above)
Comment (ejahnGithub) [reply packages/repo-mock/src/__tests__/index.test.ts:76]: make sense!

File: packages/repo-mock/src/metadata.ts (hunk 7 above)
Comment (ejahnGithub) [reply packages/repo-mock/src/metadata.ts:93]: that makes sense! I was a little bit confused by the spec..

**Pull request merged**